src/connection.rs

```diff
@@ -375,10 +371,82 @@ impl Connection {
+        // NOTE: The pubsub flag is intentionally not raised at this time since running commands
+        // within the pubsub state should not try and exit from the pubsub state.
+        PubSub::new(self)
+    }
+
+    fn exit_pubsub(&self) -> RedisResult<()> {
+        let res = self.clear_active_subscriptions();
+        if res.is_ok() {
+            self.pubsub.set(false);
+        } else {
+            // Raise the pubsub flag to indicate the connection is "stuck" in that state.
+            self.pubsub.set(true);
+        }
+
+        res
+    }
+
+    /// Get the inner connection out of a PubSub
+    ///
+    /// Any active subscriptions are unsubscribed. In the event of an error, the connection is
+    /// dropped.
+    fn clear_active_subscriptions(&self) -> RedisResult<()> {
+        // Responses to unsubscribe commands return in a 3-tuple with values
+        // ("unsubscribe" or "punsubscribe", name of subscription removed, count of remaining subs).
+        // The "count of remaining subs" includes both pattern subscriptions and non pattern
+        // subscriptions. Thus, to accurately drain all unsubscribe messages received from the
+        // server, both commands need to be executed at once.
+        {
+            // Prepare both unsubscribe commands
+            let unsubscribe = cmd("UNSUBSCRIBE").get_packed_command();
+            let punsubscribe = cmd("PUNSUBSCRIBE").get_packed_command();
+
+            // Grab a reference to the underlying connection so that we may send
+            // the commands without immediately blocking for a response.
+            let mut con = self.con.borrow_mut();
+
+            // Execute commands
+            con.send_bytes(&unsubscribe)?;
+            con.send_bytes(&punsubscribe)?;
+        }
+
+        // Receive responses
+        //
+        // There will be at minimum two responses - 1 for each of punsubscribe and unsubscribe
+        // commands. There may be more responses if there are active subscriptions. In this case,
+        // messages are received until the _subscription count_ in the responses reach zero.
+        let mut received_unsub = false;
+        let mut received_punsub = false;
+        loop {
+            let res: (Vec<u8>, (), isize) = from_redis_value(&self.recv_response()?)?;
+
+            match res.0.first().map(|v| *v) {
+                Some(b'u') => received_unsub = true,
+                Some(b'p') => received_punsub = true,
+                _ => (),
+            }
+
+            if received_unsub && received_punsub && res.2 == 0 {
+                break;
+            }
+        }
+
+        // Finally, the connection is back in its normal state since all subscriptions were
+        // cancelled *and* all unsubscribe messages were received.
+        Ok(())
+    }
 }
 
 impl ConnectionLike for Connection {
     fn req_packed_command(&self, cmd: &[u8]) -> RedisResult<Value> {
+        if self.pubsub.get() {
+            self.exit_pubsub()?;
+        }
```

Comment:
This now forces an additional check for each command sent, right? I guess the performance implications are still negligible?

Reply:
It is an additional check, but I'm guessing the branch predictor will behave favorably here. In our benchmarking of this library at OneSignal, the parser ends up being a much bigger bottleneck than anything else (we have at least 1 PR forthcoming in that regard).

Reply:
Looking forward to that PR.

Reply:
There's [a PR](https://github.com/mitsuhiko/redis-rs/pull/151) up for it as of a minute ago.